lib/Doctrine/ORM/Mapping/AssociationMapping.php

```diff
@@ -0,0 +1,310 @@
+            $mappingArray['targetEntity'],
+        );
+        unset($mappingArray['fieldName'], $mappingArray['sourceEntity'], $mappingArray['targetEntity']);
+
+        foreach ($mappingArray as $key => $value) {
+            if ($key === 'joinTable') {
+                assert($mapping instanceof ManyToManyAssociationMapping);
+
+                if ($value === [] || $value === null) {
+                    continue;
+                }
+
+                assert($mapping instanceof ManyToManyOwningSideMapping);
+
+                $mapping->joinTable = JoinTableMapping::fromMappingArray($value);
+
+                continue;
+            }
+
+            if (property_exists($mapping, $key)) {
+                $mapping->$key = $value;
+            } else {
+                throw new Exception('Unknown property ' . $key . ' on class ' . static::class);
+            }
+        }
+
+        return $mapping;
+    }
+
+    /** @psalm-assert-if-true AssociationOwningSideMapping $this */
+    final public function isOwningSide(): bool
+    {
+        return $this instanceof AssociationOwningSideMapping;
+    }
+
+    /** @psalm-assert-if-true ToOneAssociationMapping $this */
+    final public function isToOne(): bool
+    {
+        return $this instanceof ToOneAssociationMapping;
+    }
+
+    /** @psalm-assert-if-true OneToOneOwningSideMapping|ManyToOneAssociationMapping $this */
+    final public function isToOneOwningSide(): bool
+    {
+        return $this->isToOne() && $this->isOwningSide();
+    }
+
+    /** @psalm-assert-if-true ManyToManyOwningSideMapping $this */
+    final public function isManyToManyOwningSide(): bool
+    {
+        return $this instanceof ManyToManyOwningSideMapping;
+    }
+
+    final public function type(): int
+    {
+        return match (true) {
+            $this instanceof OneToOneAssociationMapping => ClassMetadata::ONE_TO_ONE,
+            $this instanceof OneToManyAssociationMapping => ClassMetadata::ONE_TO_MANY,
+            $this instanceof ManyToOneAssociationMapping => ClassMetadata::MANY_TO_ONE,
+            $this instanceof ManyToManyAssociationMapping => ClassMetadata::MANY_TO_MANY,
+            default => throw new Exception('Cannot determine type for ' . $this::class),
+        };
+    }
+
+    /** @param string $offset */
+    public function offsetExists(mixed $offset): bool
+    {
+        return isset($this->$offset) || in_array($offset, ['isOwningSide', 'type'], true);
+    }
+
+    final public function offsetGet($offset): mixed
+    {
+        return match ($offset) {
+            'isOwningSide' => $this->isOwningSide(),
+            'type' => $this->type(),
+            'isCascadeRemove' => $this->isCascadeRemove(),
+            'isCascadePersist' => $this->isCascadePersist(),
+            'isCascadeRefresh' => $this->isCascadeRefresh(),
+            'isCascadeDetach' => $this->isCascadeDetach(),
+            'isCascadeMerge' => $this->isCascadeMerge(),
+            default => $this->$offset,
```

Comment:
We should make sure the property exists.

Reply:
Should we? If we don't, we'll get the same behavior when using `$mapping['offset']` as when using `$mapping->offset`, and IMO we want that, because it means one less difference when migrating to the object API.

```
>>> $mapping['foo'];
<warning>PHP Warning:  Undefined property: Doctrine\Tests\ORM\Mapping\MyAssociationMapping::$foo in /home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php on line 213</warning>
=> null
```

Reply:
Marking as draft because adding an exception shows there are cases where I fetch properties that do not exist:

```
1) Doctrine\Tests\ORM\Functional\CompositePrimaryKeyWithAssociationsTest::testFindByAbleToGetCompositeEntitiesWithMixedTypeIdentifiers
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/CompositePrimaryKeyWithAssociationsTest.php:40

2) Doctrine\Tests\ORM\Functional\SecondLevelCacheCompositePrimaryKeyWithAssociationsTest::testFindByReturnsCachedEntity
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php:277
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheCompositePrimaryKeyWithAssociationsTest.php:47

3) Doctrine\Tests\ORM\Functional\SecondLevelCacheManyToOneTest::testPutAndLoadNonCacheableManyToOne
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php:277
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheManyToOneTest.php:195

4) Doctrine\Tests\ORM\Functional\SecondLevelCacheManyToOneTest::testPutAndLoadNonCacheableCompositeManyToOne
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php:277
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheManyToOneTest.php:233

5) Doctrine\Tests\ORM\Functional\SecondLevelCacheOneToManyTest::testPutAndLoadNonCacheableOneToMany
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php:277
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToManyTest.php:406

6) Doctrine\Tests\ORM\Functional\SecondLevelCacheOneToOneTest::testPutAndLoadNonCacheableOneToOne
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php:277
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToOneTest.php:267

7) Doctrine\Tests\ORM\Functional\Ticket\DDC117Test::testReferencesToForeignKeyEntities
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:276

8) Doctrine\Tests\ORM\Functional\Ticket\DDC117Test::testLoadManyToManyCollectionOfForeignKeyEntities
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:391
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:301

9) Doctrine\Tests\ORM\Functional\Ticket\DDC117Test::testClearManyToManyCollectionOfForeignKeyEntities
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:391
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:318

10) Doctrine\Tests\ORM\Functional\Ticket\DDC117Test::testLoadInverseManyToManyCollection
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:391
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:332

11) Doctrine\Tests\ORM\Functional\Ticket\DDC117Test::testLoadOneToManyOfSourceEntityWithAssociationIdentifier
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:391
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC117Test.php:357

12) Doctrine\Tests\ORM\Functional\Ticket\DDC2252Test::testIssue
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyOwningSideMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php:416
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php:57
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:401
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2252Test.php:75
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2252Test.php:50

13) Doctrine\Tests\ORM\Functional\Ticket\GH7062Test::testEntityWithAssociationKeyIdentityCanBeUpdated
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:661
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:686
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:236
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:1021
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:384
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php:58
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php:44

14) Doctrine\Tests\ORM\Persisters\BasicEntityPersisterCompositeTypeParametersTest::testExpandParametersWillExpandCompositeEntityKeys
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1844
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1777
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Persisters/BasicEntityPersisterCompositeTypeParametersTest.php:38

15) Doctrine\Tests\ORM\Persisters\BasicEntityPersisterCompositeTypeParametersTest::testExpandCriteriaParametersWillExpandCompositeEntityKeys
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\OneToManyAssociationMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1844
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:858
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Persisters/BasicEntityPersisterCompositeTypeParametersTest.php:52

16) Doctrine\Tests\ORM\Persisters\ManyToManyPersisterTest::testDeleteManyToManyCollection
InvalidArgumentException: Unknown property joinColumns on class Doctrine\ORM\Mapping\ManyToManyOwningSideMapping

/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Mapping/AssociationMapping.php:220
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Utility/PersisterHelper.php:73
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php:416
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php:57
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/UnitOfWork.php:401
/home/gregoire/dev/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:267
/home/gregoire/dev/doctrine-orm/tests/Doctrine/Tests/ORM/Persisters/ManyToManyPersisterTest.php:45
```

Reply:
Ah that's because of this test: https://github.com/doctrine/orm/blob/e5fb1a4a8f5aa4e37612f66a22aac10291bcbb52/lib/Doctrine/ORM/Utility/PersisterHelper.php#L78-L82

Weird how it performs the detection (instead of relying on the type)

Reply:
Fixed, as well as a similar condition below. The warnings/exceptions have disappear, please advise on https://github.com/doctrine/orm/blob/e5fb1a4a8f5aa4e37612f66a22aac10291bcbb52/lib/Doctrine/ORM/Utility/PersisterHelper.php#L78-L82, I'm convinced having that exception, even if that's temporary may be useful now :laughing:

Reply:
> Will `\InvalidArgumentException` suit you, since this is probably not meant to be caught?

I tried hard to resist my impulse to nit-pick here, but… … …

`OutOfBoundsException` was made for this.

Sorry. 😓

Reply:
Oh no, actually `OutOfRangeException` because the caller could have known in advance that the key is invalid.

Reply:
Anyway, the exception is just temporary. Any of these three would do.

🏃🏻‍♂️💨

Reply:
`OutOfRangeException` it is then… indeed, it looks perfectly suited to this 👍